File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/fragment/ChartFragment.java
package io.github.zwieback.familyfinance.business.chart.fragment;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import io.github.zwieback.familyfinance.app.FamilyFinanceApplication;
import io.github.zwieback.familyfinance.business.chart.display.ChartDisplay;
import io.github.zwieback.familyfinance.business.chart.listener.ChartDisplayListener;
import io.github.zwieback.familyfinance.business.operation.filter.OperationFilter;
import io.github.zwieback.familyfinance.business.operation.listener.OperationFilterListener;
import io.requery.Persistable;
import io.requery.reactivex.ReactiveEntityStore;

public abstract class ChartFragment<F extends OperationFilter, D extends ChartDisplay>
        extends Fragment implements OperationFilterListener<F>, ChartDisplayListener<D> {

    protected ReactiveEntityStore<Persistable> data;
    protected F filter;
    protected D display;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        data = ((FamilyFinanceApplication) ((Activity) extractContext()).getApplication()).getData();
        filter = loadFilter(savedInstanceState);
        display = loadDisplay(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(getFilterName(), filter);
        outState.putParcelable(getDisplayName(), display);
    }

    private F loadFilter(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            return createDefaultFilter();
        }
        return savedInstanceState.getParcelable(getFilterName());
    }

    private D loadDisplay(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            return createDefaultDisplay();
        }
        return savedInstanceState.getParcelable(getDisplayName());
    }

    protected abstract String getFilterName();

    protected abstract F createDefaultFilter();

    protected abstract String getDisplayName();

    protected abstract D createDefaultDisplay();

    public abstract void showFilterDialog();

    public abstract void showDisplayDialog();

    @NonNull
    protected Context extractContext() {
        Context context = super.getContext();
        if (context != null) {
            return context;
        }
        throw new IllegalStateException("Context is null");
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/fragment/ChartFragment.kt
package io.github.zwieback.familyfinance.business.chart.fragment

import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import com.github.mikephil.charting.charts.Chart
import com.github.mikephil.charting.data.Entry
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.app.FamilyFinanceApplication
import io.github.zwieback.familyfinance.business.chart.display.ChartDisplay
import io.github.zwieback.familyfinance.business.chart.listener.ChartDisplayListener
import io.github.zwieback.familyfinance.business.chart.service.converter.OperationConverter
import io.github.zwieback.familyfinance.business.chart.service.grouper.OperationGrouper
import io.github.zwieback.familyfinance.business.operation.filter.OperationFilter
import io.github.zwieback.familyfinance.business.operation.listener.OperationFilterListener
import io.github.zwieback.familyfinance.core.model.OperationView
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.requery.Persistable
import io.requery.query.Result
import io.requery.reactivex.ReactiveEntityStore

abstract class ChartFragment<CHART, ENTRY, FILTER, DISPLAY> :
    Fragment(),
    OperationFilterListener<FILTER>,
    ChartDisplayListener<DISPLAY>
        where CHART : Chart<*>,
              ENTRY : Entry,
              FILTER : OperationFilter,
              DISPLAY : ChartDisplay<DISPLAY> {

    protected lateinit var data: ReactiveEntityStore<Persistable>
    protected lateinit var filter: FILTER
    protected lateinit var display: DISPLAY
    protected lateinit var chart: CHART

    protected lateinit var operationConverter: OperationConverter<ENTRY>
    protected lateinit var operationGrouper: OperationGrouper

    private val compositeDisposable = CompositeDisposable()

    var isDataLoaded: Boolean = false
        private set

    protected abstract val filterName: String

    protected abstract val displayName: String

    @get:LayoutRes
    protected abstract val fragmentChartLayout: Int

    @get:IdRes
    protected abstract val chartId: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        data = ((requireContext() as Activity).application as FamilyFinanceApplication).data
        filter = loadFilter(savedInstanceState)
        display = loadDisplay(savedInstanceState)

        operationConverter = determineOperationConverter()
        operationGrouper = determineOperationGrouper()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(fragmentChartLayout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        chart = view.findViewById(chartId)
        setupChart()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(filterName, filter)
        outState.putParcelable(displayName, display)
    }

    override fun onDestroy() {
        compositeDisposable.clear()
        super.onDestroy()
    }

    private fun loadFilter(savedInstanceState: Bundle?): FILTER {
        return savedInstanceState?.getParcelable(filterName) ?: createDefaultFilter()
    }

    private fun loadDisplay(savedInstanceState: Bundle?): DISPLAY {
        return savedInstanceState?.getParcelable(displayName) ?: createDefaultDisplay()
    }

    protected abstract fun createDefaultFilter(): FILTER

    protected abstract fun createDefaultDisplay(): DISPLAY

    protected abstract fun setupChart()

    fun refreshData() {
        isDataLoaded = true
        clearData(R.string.chart_loading)

        compositeDisposable.add(
            Single.fromCallable { buildOperations() }
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .map { operations -> groupOperations(operations) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { groupedOperations -> showData(groupedOperations) }
        )
    }

    protected fun clearData(@StringRes noDataTextRes: Int) {
        chart.setNoDataText(getString(noDataTextRes))
        chart.clear()
    }

    protected abstract fun buildOperations(): Result<OperationView>

    private fun groupOperations(operations: Result<OperationView>): Map<Float, List<OperationView>> {
        return operationGrouper.group(operations.toList(), filter.startDate, filter.endDate)
    }

    protected fun convertOperations(groupedOperations: Map<Float, List<OperationView>>): List<ENTRY> {
        return operationConverter.convertToEntries(groupedOperations)
    }

    protected abstract fun showData(groupedOperations: Map<Float, List<OperationView>>)

    abstract fun showFilterDialog()

    abstract fun showDisplayDialog()

    protected abstract fun determineOperationConverter(): OperationConverter<ENTRY>

    protected abstract fun determineOperationGrouper(): OperationGrouper
}
